#' High-Throughput Genome-wide Translocation Sequencing Analysis
#'
#' This function is used to do a High-Throughput Genome-wide Translocation Sequencing analysis,
#' it returns the output data in the specific folder output_dir_name, found inside the input_dir_path
#'
#' @param input_dir_path, a character string indicating the path of a directory containing all the input data
#' @param fastq1_name, a character string indicating the first input FASTQ file name
#' @param fastq2_name, a character string indicating the second input FASTQ file name (optional)
#' @param expInfo_name, a character string indicating name of the libseqInfo.txt file generated by generateLibInfo.sh
#' @param expInfo2_name, a character string indicating name of the libseqInfo2.txt file generated by generateLibInfo.sh (optional)
#' @param output_dir_name, a character string indicating the name of the output folder (which has to be in the input_dir_path folder)
#' @param configType, a character string indicating the configuration type, as defined in config_types.json. Examples include "HTGTS_mouse", "HTGTS_human", "CELTICSseq", "polyA"
#' @param assembly, a character string indicating the reference genome version (e.g., mm9)
#' @author Luca Alessandr√¨, Agata D'Onofrio
#'
#' @examples
#' \dontrun{
#'  htgts(
#'     input_dir_path = "/the/input/path",
#'     fastq1_name = "filtered_JT8572_S1_R1_001.fastq",
#'     fastq2_name = "filtered_JT8572_S1_R2_001.fastq",
#'     expInfo_name = "libseqInfo.txt",
#'     expInfo2_name = "libseqInfo2.txt",
#'     output_dir_name = "output_directory",
#'     configType = "HTGTS_mouse",
#'     assembly = "mm9"
#'  )
#' }
#' @export
htgts <- function(input_dir_path, fastq1_name, fastq2_name = NULL, expInfo_name, expInfo2_name = NULL, output_dir_name, configType, assembly){

  # Type checking.
  if (typeof(input_dir_path) != "character") {
    stop(
      paste(
        "input_dir_path type is", paste0(typeof(input_dir_path), "."),
        "It should be \"character\""
      )
    )
  }
  if (typeof(fastq1_name) != "character") {
    stop(
      paste(
        "fastq1_name type is", paste0(typeof(fastq1_name), "."),
        "It should be \"character\""
      )
    )
  }
  if (!is.null(fastq2_name) && typeof(fastq2_name) != "character") {
    stop(
      paste(
        "fastq2_name type is", paste0(typeof(fastq2_name), "."),
        "It should be \"character\""
      )
    )
  }
  if (typeof(expInfo_name) != "character") {
    stop(
      paste(
        "expInfo_name type is", paste0(typeof(expInfo_name), "."),
        "It should be \"character\""
      )
    )
  }
  if (!is.null(expInfo2_name) && typeof(expInfo2_name) != "character") {
    stop(
      paste(
        "expInfo2_name type is", paste0(typeof(expInfo2_name), "."),
        "It should be \"character\""
      )
    )
  }
  if (typeof(output_dir_name) != "character") {
    stop(
      paste(
        "output_dir_name type is", paste0(typeof(output_dir_name), "."),
        "It should be \"character\""
      )
    )
  }
  if (typeof(configType) != "character") {
    stop(
      paste(
        "configType type is", paste0(typeof(configType), "."),
        "It should be \"character\""
      )
    )
  }
  if (typeof(assembly) != "character") {
    stop(
      paste(
        "assembly type is", paste0(typeof(assembly), "."),
        "It should be \"character\""
      )
    )
  }

  # Check if input_dir_path exists
  if (!is_running_in_docker()) {
    if (!dir.exists(input_dir_path)) {
      stop(paste("input_dir_path:", input_dir_path, "does not exist."))
  }
  if (!dir.exists(paste0(input_dir_path, "/", output_dir_name))) {
      stop(paste("output_dir_path:", paste0(input_dir_path, "/", output_dir_name), "does not exist."))
    }
  }

  # Checking if a second fastq file was provided
  if (!is.null(fastq2_name)) {
      additional_arguments = c(
      "/Algorithm/HTGTS_Full.sh",
        paste0("-fastq1 /Data/", fastq1_name),
        paste0("-fastq2 /Data/", fastq2_name),
        paste0("-expInfo /Data/", expInfo_name),
        paste0("-expInfo2 /Data/", expInfo2_name),
        paste0("-outDir /Data/", output_dir_name),
        paste("-configType", configType),
        paste("-assembly", assembly)
        )
      }
    else {
      additional_arguments = c(
      "/Algorithm/HTGTS_Full.sh",
        paste0("-fastq1 /Data/", fastq1_name),
        paste0("-expInfo /Data/", expInfo_name),
        paste0("-outDir /Data/", output_dir_name),
        paste("-configType", configType),
        paste ("-assembly", assembly)
        )
      }

  # Executing the docker job
  run_in_docker(
    image_name = paste0("repbioinfo/htgts_pipeline_lts_v16:latest"),
    volumes = list(
      c(input_dir_path, "/Data")
    ),
    additional_arguments
  )
}

